pkg/signature/kms/cliplugin/common/interface.go

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright 2021 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package common defines the JSON schema for plugin arguments and return values.
+package common
+
+import (
+	"crypto"
+	"time"
+)
+
+const (
+	// ProtocolVersion is the version of the schema and communication protocol for the plugin system.
+	// Breaking changes to the PluginClient and this schema necessarily mean major version bumps of
+	// this ProtocolVersion and the sigstore version.
+	// Plugin authors may choose to be backwards compatible with older versions.
+	ProtocolVersion            = "1"
+	DefaultAlgorithmMethodName = "defaultAlgorithm"
+	CreateKeyMethodName        = "createKey"
+	// TODO: Additonal methods to be implemented
+)
+
+// PluginArgs contains all the initialization and method arguments to be sent to the plugin as a CLI argument.
+type PluginArgs struct {
+	*MethodArgs
+	InitOptions *InitOptions `json:"initOptions"`
+}
+
+// InitOptions contains the initial arguments when calling cliplugin.LoadSignerVerifier().
+type InitOptions struct {
+	ProtocolVersion string      `json:"protocolVersion"`
+	KeyResourceID   string      `json:"keyResourceID"`
+	HashFunc        crypto.Hash `json:"hashFunc"`
+	// TODO: extracted values from signature.RPCOption from LoadSignerVerifier().
+}
+
+// MethodArgs contains the method arguments. MethodName must be specified,
+// while any one of the other fields describing method arguments must also be specified.
+type MethodArgs struct {
+	// MethodName specifies which method is intended to be called.
+	MethodName       string                `json:"methodName"`
+	DefaultAlgorithm *DefaultAlgorithmArgs `json:"defaultAlgorithm,omitempty"`
+	CreateKey        *CreateKeyArgs        `json:"createKey,omitempty"`
+	// TODO: Additonal methods to be implemented
+}
+
+// PluginResp contains the serialized plugin method return values.
+type PluginResp struct {
+	ErrorMessage     string                `json:"errorMessage,omitempty"`
+	DefaultAlgorithm *DefaultAlgorithmResp `json:"defaultAlgorithm,omitempty"`
+	CreateKey        *CreateKeyResp        `json:"createKey,omitempty"`
+	// TODO: Additonal methods to be implemented
+}
+
+type DefaultAlgorithmArgs struct {
+}
+
+type DefaultAlgorithmResp struct {
+	DefaultAlgorithm string
+}
+
+type CreateKeyArgs struct {
+	CtxDeadline *time.Time
```

Comment:
Could this come from InitOptions rather than having to define this as a parameter for every method?

Reply:
I might be misunderstanding this. I should this would be for the timeout for each call from the client to the plugin. If it's long-running, that should be fine as each timeout is computed for that invocation.

In the spirit of simplifying what plugin authors need to create, I think keeping these methods small in scope would be good. If we later need to add per-method timeouts, that's a non-breaking change.

Reply:
The CtxDeadline is not a duration, it’s a specific timestamp in the future. But I think we should fully implement the interface, and plugin authors can choose to ignore the CtxDeadline. Our CommandContext invocation will respect a CtxDeadline, if the caller provides one.

Reply:
But since the plugin Programs themselves are not long-running, and author can interpret InitOptions.CyxDeadline, as a deadline for doing cold-start initialization tasks, like fetching a TUF cache, for example. 

I’ll add a TODO to add some guidance about handling Deadlines and InitOptions.

Reply:
Sounds good - related to the other comment, let's document what is a must for handling and what is optional.